### PR TITLE
AzureApiManagement new entity "Tag" documented

### PR DESCRIPTION
**Pull request by vfedonkin**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [ * ] The title of the PR is clear and informative.
- [ *  ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ * ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ * ] If applicable, the PR references the bug/issue that it fixes.
- [ * ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ * ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ * ] My spec meets the review criteria:
  - [ * ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ * ] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ * ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by vfedonkin** [comment]
@solankisamir (Samir Solanski) , could you please also review this PR?

**Comment by solankisamir** [comment]
Update the readme.md file with the new resource files for tags.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+          "$ref": "#/definitions/TagContractProperties",
+          "description": "Tag entity contract properties."
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "./apimanagement.json#/definitions/Resource"
+        }
+      ],
+      "description": "Contract details."
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: >Contract [](start = 22, length = 8)

Tag Contract Details.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+          "description": "Description of the Tag."
+        },
+        "externalDocsUrl": {
+          "type": "string",
+          "description": "Absolute URL of external resources describing the tag."
+        },
+        "externalDocsDescription": {
+          "type": "string",
+          "description": "Description of the external resources describing the tag."
+        }
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: you can put these parameters in a base contract and do allOf in CreateParameterProperties and ContractProperties. would reduce redundancy

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        "product": {
+          "$ref": "./apimproducts.json#/definitions/ProductContract",
+          "description": "Product associated with the tag."
+        }
+      },
+      "description": "TagResource contract properties."
+    },
+    "OperationEntityContract": {
+      "properties": {
+        "displayName": {
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json]: is there any max/min length to any of the properties in the contract.

Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json]: This is read-only contract, we never use in any put/patch requests. Should we specify min/max length anyway?

Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json]: then we can specify "readonly": "true"

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json (hunk 1 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: Done

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json (hunk 2 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: Done. Also got rid of some unnecessary duplicates in TagContract

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json (hunk 3 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json]: Done

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+          "description": "Properties supplied to Create Tag operation."
+        }
+      },
+      "description": "Parameters supplied to the Create Tag operation."
+    },
+    "TagUpdateParameters": {
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/TagContractProperties",
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: is displayName a required property in the UpdateContract. if yes... then we dont need 2 contracts. you can just have one contract for both Create and Update.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+          "type": "string",
+          "description": "Description of the Tag."
+        },
+        "externalDocsUrl": {
+          "type": "string",
+          "description": "Absolute URL of external resources describing the tag."
+        },
+        "externalDocsDescription": {
+          "type": "string",
+          "description": "Description of the external resources describing the tag."
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: is there a max min to these properties also?

**Review by solankisamir — APPROVED** [review]
left some minor comments. rest looks good. Thanks @vfedonkin

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 6]
```diff
@@ -914,7 +914,282 @@
...
+            "$ref": "./apimanagement.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "./apimanagement.json#/parameters/ServiceNameParameter"
+          },
+          {
+            "$ref": "#/parameters/ProductIdParameter"
+          },
+          {
+            "$ref": "#./apimtags.json#/parameters/TagIdParameter"
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: >#./ [](start = 21, length = 3)

maybe this is what the CI job is complaining about.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 7]
```diff
@@ -914,7 +914,282 @@
...
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: >#./ [](start = 21, length = 3)

check this

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 8]
```diff
@@ -914,7 +914,282 @@
...
+          },
+          {
+            "$ref": "./apimanagement.json#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Gets the details of the tag specified by its identifier.",
+            "schema": {
+              "$ref": "#./apimtags.json#/definitions/TagContract"
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: >#./ [](start = 23, length = 3)

check this.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json [hunk 9]
```diff
@@ -851,16 +851,873 @@
...
+            "$ref": "./apimanagement.json#/parameters/ServiceNameParameter"
+          },
+          {
+            "$ref": "#/parameters/ApiIdParameter"
+          },
+          {
+            "$ref": "#/parameters/OperationIdParameter"
+          },
+          {
+            "$ref": "#./apimtags.json#/parameters/TagIdParameter"
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json]: >#./ [](start = 21, length = 3)

check this and other references below.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json [hunk 10]
```diff
@@ -935,6 +1792,9 @@
           {
             "$ref": "#/parameters/ApiIdParameter"
           },
+          {
+            "$ref": "#./apimtags.json#/parameters/TagIdParameter"
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json]: >#./ [](start = 21, length = 3)

you can fix this and run the tool [https://github.com/Azure/oav](https://github.com/Azure/oav) locally before submitting again.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json (hunk 5 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtags.json]: Yes, we need for externalDocsUrl only. Added.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 6 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: fixed

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 7 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: fixed

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 8 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: fixed

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json (hunk 9 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json]: fixed

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json (hunk 10 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json]: fixed

**Comment by fearthecowboy** [comment]
Hey @solankisamir / vfedonkin

Can y'all ping me when you're ready for me to review this?

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 11]
```diff
@@ -914,7 +918,282 @@
           }
         }
       }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/products/{productId}/tags": {
+      "get": {
+        "tags": [
+          "ProductTags"
+        ],
+        "operationId": "Tags_ListByProduct",
```
Comment (fearthecowboy) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: This operation is `Tags_ListByProduct` the next is `Tag_HeadByProduct`

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 12]
```diff
@@ -914,7 +918,282 @@
...
+              "$ref": "./apimanagement.json#/definitions/ErrorResponse"
+            }
+          }
+        }
+      },
+      "delete": {
+        "tags": [
+          "ProductTags"
+        ],
+        "operationId": "Tag_DeleteFromProduct",
```
Comment (fearthecowboy) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: I'd recommend this was just `Tags_Delete`  -- the operation Id is used to generate method names, and unless there is an overriding or compelling reason, we'd like to stick to common patterns.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 13]
```diff
@@ -914,7 +918,282 @@
...
+              "$ref": "./apimanagement.json#/definitions/ErrorResponse"
+            }
+          }
+        }
+      },
+      "put": {
+        "tags": [
+          "ProductTags"
+        ],
+        "operationId": "Tag_CreateOrUpdateForProduct",
```
Comment (fearthecowboy) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: I'd recommend this was just `Tags_CreateOrUpdate` -- the operation Id is used to generate method names, and unless there is an overriding or compelling reason, we'd like to stick to common patterns.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 11 above)
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: We have used a singular case in all other operations. @vfedonkin let's use Tag.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 14]
```diff
@@ -914,7 +918,282 @@
...
+              "$ref": "./apimanagement.json#/definitions/ErrorResponse"
+            }
+          }
+        }
+      },
+      "get": {
+        "tags": [
+          "ProductTags"
+        ],
+        "operationId": "Tag_GetByProduct",
```
Comment (fearthecowboy) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: I'd recommend this was just `Tags_Get` -- the operation Id is used to generate method names, and unless there is an overriding or compelling reason, we'd like to stick to common patterns.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 15]
```diff
@@ -914,7 +918,282 @@
...
+        },
+        "x-ms-odata": "./apimtags.json#/definitions/TagContract"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/products/{productId}/tags/{tagId}": {
+      "head": {
+        "tags": [
+          "ProductTags"
+        ],
+        "operationId": "Tag_HeadByProduct",
```
Comment (fearthecowboy) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: We don't see a lot of `HEAD` operations.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 14 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: @fearthecowboy 
We already have Tag_Get in apimtags.json with url like 
".../service/{serviceName}/tags/{tagId}"
Can we use the same operationId for
".../service/{serviceName}/products/{productId}/tags/{tagId}" ?

**Comment by fearthecowboy** [comment]
@vfedonkin 

> We already have Tag_Get in apimtags.json with url like
>".../service/{serviceName}/tags/{tagId}"
>Can we use the same operationId for
>".../service/{serviceName}/products/{productId}/tags/{tagId}" ?

Ah, I'm just seeing that. 

HMMMM.

No, we definitely can't have duplicate `operationId`s.

In that case, this could remain the way it is.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 11 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: Done, using "Tag" everywhere.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 15 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: Done. The same for apis/operations/tags entities.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 13 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: Renamed to Tag_AssignToProduct. Same for Tag_AssignToApi and Tag_AssignToOperation

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json (hunk 12 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: Renamed to Tag_DetachFromProduct. Same for Tag_DetachFromApi and Tag_DetachFromOperation

**Comment by fearthecowboy** [comment]
Adding @ravbhatnagar   -- new APIs require ARM signoff.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json [hunk 16]
```diff
@@ -851,16 +851,873 @@
...
+        },
+        "x-ms-odata": "./apimtags.json#/definitions/TagContract"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}/operations/{operationId}/tags/{tagId}": {
+      "head": {
+        "tags": [
+          "OperationTags"
+        ],
+        "operationId": "Tag_HeadByOperation",
```
Comment (fearthecowboy) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json]: Perhaps `Tag_GetEntityStateByOperation` ?

**Review by fearthecowboy — COMMENTED** [review]
Just a couple of things; nothing serious.

We should put a suppression in the `readme.md` for the two linter violations; add a section that looks like this:

~~~
# Suppressions

``` yaml
directive: 
  - suppress: R3016
     reason: existing properties, can't be changed without breaking API.
```
~~~

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json (hunk 16 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimapis.json]: Done

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/tagResources": {
+      "get": {
+        "tags": [
+          "TagResources"
+        ],
+        "operationId": "TagResources_ListByService",
```
[the remark this answers is not in the document]
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimtagresources.json]: Done

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json [hunk 18]
```diff
@@ -914,7 +918,282 @@
...
+            }
+          }
+        }
+      },
+      "delete": {
+        "tags": [
+          "ProductTags"
+        ],
+        "operationId": "Tag_DeleteFromProduct",
+        "description": "Detach the tag from the Product.",
```
[the remark this answers is not in the document]
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: Tag is really exists as an entity, it's just detached from the product.

[the remark this answers is not in the document]
Comment (fearthecowboy) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2017-03-01/apimproducts.json]: Oh. interesting.

**Review by fearthecowboy — APPROVED** [review]
I'm good with the state of this. Awaiting ARM signoff from @ravbhatnagar

**Comment by vfedonkin** [comment]
Hi @ravbhatnagar, do I need to do anything else or just wait until it merged?
Thanks.

**Comment by solankisamir** [comment]
@ravbhatnagar can you take a look?

**Comment by ravbhatnagar** [comment]
Signing off!

**Comment by fearthecowboy** [comment]
@vfedonkin We're done here -- are we ok to merge?

**Comment by vfedonkin** [comment]
@fearthecowboy , yes, but I need to resolve conflicts first. Will let you know when it's done.

**Comment by vfedonkin** [comment]
@fearthecowboy, conflicts solved, ok to merge.

**Comment by fearthecowboy** [comment]
You have some validation issues : https://travis-ci.org/Azure/azure-rest-api-specs/jobs/335311496

**Comment by vfedonkin** [comment]
@fearthecowboy, validation issues are not related to my changes but I will investigate what's wrong.

**Comment by vfedonkin** [comment]
Moved all tags-related paths to apimtags.json file since apimapis.json is already too huge.

**Comment by vfedonkin** [comment]
@fearthecowboy, could you please check now? I left apimapis.json file as it was before, without any changes.

**Comment by solankisamir** [comment]
@fearthecowboy a gentle ping on this one.

**Comment by vfedonkin** [comment]
@fearthecowboy , we couldn't fix this issue, as we understand it's an issue with validation tool itself. 
It doesn't support boolean parameters in examples.

Below is a parameter definition:
          {
            "name": "deleteSubscriptions",
            "in": "query",
            "required": false,
            "type": "boolean",
            "description": "Delete existing subscriptions to the product or not."
          },

And this is a sample file:
{
    "parameters": {
        "serviceName": "apimService1",
        "resourceGroupName": "rg1",
        "api-version": "2017-03-01",
        "subscriptionId": "subid",
        "productId": "testproduct",
        "deleteSubscriptions": "true",
        "If-Match": "*"
    },
    "responses": {
        "204": {}
    }
}

What is the best way to fix it?
Thanks,
Vitaliy.

**Comment by fearthecowboy** [comment]
`true` should not be in quotes. `"true" is not a boolean value`

**Comment by vfedonkin** [comment]
We tried, it didn't help. 
We have exactly the same error when we run this test tool locally:
oav validate-example <...>\specification\apimanagement\resource-manager\Microsoft.ApiManagement\stable\2017-03-01\apimproducts.json

**Comment by fearthecowboy** [comment]
@vfedonkin  - I'll try doing this locally and grab @amarzavery  if it's not workable.

**Comment by solankisamir** [comment]
@fearthecowboy any update on this.

**Comment by fearthecowboy** [comment]
@vfedonkin  - It was exactly as I said; the quotes around `"true"` was causing it to fail. I fixed it in your branch for you.

**Pull request merged**